rosbag2_cpp/include/rosbag2_cpp/writers/buffer_layer.hpp

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020, Robotec.ai sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__WRITERS__BUFFER_LAYER_HPP_
+#define ROSBAG2_CPP__WRITERS__BUFFER_LAYER_HPP_
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#include "rosbag2_storage/storage_interfaces/read_write_interface.hpp"
+#include "rosbag2_storage/storage_options.hpp"
+#include "rosbag2_cpp/visibility_control.hpp"
+
+namespace rosbag2_cpp
+{
+namespace writers
+{
+
+class ROSBAG2_CPP_PUBLIC BufferLayer
+{
+public:
+  BufferLayer(
+    std::shared_ptr<rosbag2_storage::storage_interfaces::ReadWriteInterface> storage,
+    const uint64_t & max_buffer_size);
+  ~BufferLayer();
+
+  // Flush data into storage, and reset cache
+  void reset_cache();
+  // Push data into primary buffer
+  bool push(std::shared_ptr<const rosbag2_storage::SerializedBagMessage> msg);
+  // Flush buffers and reset them, shut down consumer thread
+  void close();
+  // Starts consumer thread if one is not working
+  void start_consumer();
```

Comment:
I would appreciate if we could separate the storage in the buffer from the execution model. That way we can have the buffer being a single task class whereas the threading can be handled in various ways independently of the buffer.